lib/Parse/ParseType.cpp

```diff
@@ -237,22 +237,39 @@ ParserResult<TypeRepr> Parser::parseType(Diag<> MessageID,
   return ty;
 }
 
-ParserResult<TypeRepr> Parser::parseTypeIdentifierWithRecovery(
+ParserResult<TypeRepr> Parser::parseTypeForInheritance(
     Diag<> MessageID, Diag<TypeLoc> NonIdentifierTypeMessageID) {
-  ParserResult<TypeRepr> Ty = parseType(MessageID);
-
-  if (!Ty.isParseError() && !isa<IdentTypeRepr>(Ty.get()) &&
-      !isa<ErrorTypeRepr>(Ty.get())) {
-    diagnose(Ty.get()->getStartLoc(), NonIdentifierTypeMessageID, Ty.get())
-        .highlight(Ty.get()->getSourceRange());
-    Ty.setIsParseError();
-    Ty = makeParserResult(
-        Ty, new (Context) ErrorTypeRepr(Ty.get()->getSourceRange()));
+  ParserResult<TypeRepr> Ty = parseTypeSimpleOrComposition(MessageID);
+
+  if (Ty.hasCodeCompletion())
+    return makeParserCodeCompletionResult<TypeRepr>();
+
+  if (Ty.isParseError() || isa<ErrorTypeRepr>(Ty.get()))
+    return Ty;
+
+  if (isa<IdentTypeRepr>(Ty.get()))
+    return Ty;
+
+  if (auto Comp = dyn_cast<CompositionTypeRepr>(Ty.get())) {
+    bool hasNonIdent = false;
+    for (auto T : Comp->getTypes()) {
+      if (isa<IdentTypeRepr>(T))
+        continue;
+      hasNonIdent = true;
+      diagnose(T->getLoc(), NonIdentifierTypeMessageID, T)
+        .highlight(T->getSourceRange());
+    }
+    // IdentType only composition are allowed.
```

Comment:
Just to be clear, is this a behavior change? Is anything additional being accepted that wasn't before, and is anything additional being *rejected* that wasn't before?